commands/service_publish.go

```diff
@@ -0,0 +1,99 @@
+package commands
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+
+	"github.com/docker/docker/pkg/archive"
+	"github.com/mesg-foundation/core/ipfs"
+	"github.com/mesg-foundation/core/service/importer"
+	"github.com/mesg-foundation/core/utils/readme"
+	"github.com/spf13/cobra"
+)
+
+// PublishVersion is the version used to publish the services to the marketplace
+const PublishVersion = 1
+
+type servicePublishCmd struct {
+	baseCmd
+
+	path string
+	ipfs *ipfs.IPFS
+
+	e ServiceExecutor
+}
+
+type marketplaceData struct {
+	Version int `json:"version"`
+	Service struct {
+		Deployment struct {
+			Source string `json:"source"`
+		} `json:"deployment"`
+	} `json:"service"`
+	Definition *importer.ServiceDefinition `json:"definition"`
+	Readme     string                      `json:"readme,omitempty"`
+}
+
+func newServicePublishCmd(e ServiceExecutor) *servicePublishCmd {
+	c := &servicePublishCmd{e: e}
+	c.cmd = newCommand(&cobra.Command{
+		Use:     "publish",
+		Short:   "Publish a service on the MESG Marketplace",
+		Example: `mesg-core service publish PATH_TO_SERVICE`,
+		PreRunE: c.preRunE,
+		RunE:    c.runE,
+		Args:    cobra.MaximumNArgs(1),
+		Hidden:  true, // TODO: Remove when this feature is finished
+	})
+	return c
+}
+
+func (c *servicePublishCmd) preRunE(cmd *cobra.Command, args []string) error {
+	c.path = getFirstOrDefault(args)
+	c.ipfs = ipfs.New()
+	return nil
+}
+
+func (c *servicePublishCmd) runE(cmd *cobra.Command, args []string) error {
+	tar, err := archive.TarWithOptions(c.path, &archive.TarOptions{
+		Compression: archive.Gzip,
+	})
+	if err != nil {
+		return err
+	}
+	tarballResponse, err := c.ipfs.Add("tarball", tar)
+	if err != nil {
+		return err
+	}
+
+	definitionFile, err := c.createDefinitionFile(tarballResponse.Hash)
+	if err != nil {
+		return err
+	}
+
+	definitionResponse, err := c.ipfs.Add("definition", bytes.NewReader(definitionFile))
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("https://gateway.ipfs.io/ipfs/" + definitionResponse.Hash)
+
+	return nil
+}
+
+func (c *servicePublishCmd) createDefinitionFile(tarballHash string) ([]byte, error) {
+	definition, err := importer.From(c.path)
+	if err != nil {
+		return nil, err
+	}
+	var data marketplaceData
+	data.Version = PublishVersion
+	data.Service.Deployment.Source = tarballHash
```

Comment:
What is the value of `tarballHash`?
Just the hash? Does it contains an `ipfs://` prefix? Is it an http url?

Reply:
just the hash

Reply:
I think it should be an URL with the ipfs prefix in this case.
We don't want to have to guess that's a ipfs hash, we want to be sure.
And like this the marketplace can be compatible with any file storage network.

Reply:
I didn't find any information about an IPFS protocol. We could consider this as a protocol but i  don't like the idea of supporting something that doesn't exists.

We could have a deployment type attribute or put the full address of gateway ipfs, eg: https://gateway.ipfs.io/ipfs/xxxx

Reply:
Yeah right....
The best "official" answer:
> IPFS data is designed to be linked in canonical path (without schema) but support for it is not yet in browsers. In case of storage and internal representation I would use this format: `/ipfs/QmburSK2Cjav26X6oGsiXwwwDXnVUUjm9i65UErRD99Z9v`
> From https://github.com/ipfs/ipfs/issues/181

We need to have something that say that's an IPFS hash. Otherwise we are stuck with IPFS for link on the marketplace..

Reply:
or we can anyway add a type for this, in this case it will be type ipfs and we could have type url, type git... instead of having to do a detection on the url we can explicitly say it

Reply:
deployment type added